File: swift/Sources/GildedRose/ItemCategory/Conjured.swift
//
//  Conjured.swift
//  
//
//  Created by Frank Thamel on 2022-08-19.
//

import Foundation

class Conjured: Item {
    override func updateQuality() {
        degradeQuality()
        sellIn = sellIn - 1
    
        guard sellIn < 0 else { return }
        
        degradeQuality()
    }
    
    private func degradeQuality() {
        guard quality > 1 else {
            quality = 0
            return
        }
 
        quality = quality - 2
    }
}

File: ruby/update_handlers/conjured.rb
# frozen_string_literal: true

require_relative 'base'

module UpdateHandlers
  class Conjured < Base
    private

    def update_sell_in
      item.sell_in -= 1
    end

    def update_quality
      quality_value = sell_in.negative? ? quality - 4 : quality - 2

      item.quality = normalise(quality_value)
    end
  end
end
